### PR TITLE
[llava][15/N] Extract out text decoder runner

### PR DESCRIPTION
**Pull request by larryliu0820**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #4588
* #4585
* __->__ #4567
* #4556

Last PR #4556 refactored run_model_step() so that it is suitable to be
extracted out as a separate class.

This new `TextDecoderRunner` provides 2 APIs:

* step(tokens, start_pos)
This API takes one or more tokens with start_pos and feed them into Module. Return a tensor of logits.

* logits_to_token(logits)
This API samples the result and returns a token. We don't expect this
logic to change across different runners.

Differential Revision: [D60856571](https://our.internmc.facebook.com/intern/diff/D60856571)

### CODE REVIEW

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60856571).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60856571).

**Review by JacobSzwejbka — APPROVED** [review]

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60856571).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60856571).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60856571).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60856571).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60856571).

**Comment by larryliu0820** [comment]
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60856571).
